standardize entr_detr closures

haakon-e opened the pull request.

charleskawczynski commented on src/closures/entr_detr.jl:
```diff
@@ -109,6 +104,38 @@ function entr_detr(param_set::APS, εδ_model_vars, εδ_model_type::MDEntr)
+    Δb = εδ_model_vars.b_up - εδ_model_vars.b_en
+    μ_ij = (χ - εδ_model_vars.a_up / (εδ_model_vars.a_up + εδ_model_vars.a_en)) * Δb / Δw
+    exp_arg = μ_ij / μ_0
+    D_ε = 1.0 / (1.0 + exp(-exp_arg))
+    D_δ = 1.0 / (1.0 + exp(exp_arg))
```
```suggestion
    D_ε = 1 / (1 + exp(-exp_arg))
    D_δ = 1 / (1 + exp(exp_arg))
```

charleskawczynski commented on src/closures/entr_detr.jl:
```diff
@@ -109,6 +104,38 @@ function entr_detr(param_set::APS, εδ_model_vars, εδ_model_type::MDEntr)
+    D_ε = 1.0 / (1.0 + exp(-exp_arg))
+    D_δ = 1.0 / (1.0 + exp(exp_arg))
+
+    M_δ = (max((εδ_model_vars.RH_up)^β - (εδ_model_vars.RH_en)^β, 0.0))^(1.0 / β)
+    M_ε = (max((εδ_model_vars.RH_en)^β - (εδ_model_vars.RH_up)^β, 0.0))^(1.0 / β)
```
```suggestion
    M_δ = (max((εδ_model_vars.RH_up)^β - (εδ_model_vars.RH_en)^β, 0))^(1 / β)
    M_ε = (max((εδ_model_vars.RH_en)^β - (εδ_model_vars.RH_up)^β, 0))^(1 / β)
```

charleskawczynski reviewed (approved):
So, if I understand correctly, this narrows the scope of the function we want to hack? We only need to dispatch on `non_dimensional_part`?

I added some suggestions to fix type stability, but LGTM aside from that

yairchn commented on src/closures/entr_detr.jl:
```diff
@@ -109,6 +104,38 @@ function entr_detr(param_set::APS, εδ_model_vars, εδ_model_type::MDEntr)
+ - `εδ_model_type`  :: MDEntr - Moisture deficit entrainment closure
+"""
+function non_dimensional_part(param_set, εδ_model_vars, ::MDEntr)
+    Δw = get_Δw(param_set, εδ_model_vars)
```
you need 'FT = eltype(εδ_model_vars)' here

costachris commented on src/closures/entr_detr.jl:
```diff
@@ -109,6 +104,39 @@ function entr_detr(param_set::APS, εδ_model_vars, εδ_model_type::MDEntr)
 end
 
+"""
+    entr_detr(param_set, εδ_model_vars, εδ_model_type)
```
I think we can get rid of this doc string for `entr_detr` and the one below, or redefine it for `non_dimensional_part`

costachris commented on src/closures/entr_detr.jl:
```diff
@@ -119,10 +147,8 @@ network to predict the non-dimensional component of dynamical entrainment:
-    area_limiter = max_area_limiter(param_set, εδ_model_vars)
 
     # Neural network closure
     nn_arc = (4, 2, 2)  # (#inputs, #neurons, #outputs)
```
I had to make the following change to the activation functions definitions since `flux` is now imported in src/TurbulenceConvection.jl

`sigmoid` -> `Flux.sigmoid`
`softplus` -> `Flux.softplus`

charleskawczynski replied:
Good catch!

haakon-e commented:
> So, if I understand correctly, this narrows the scope of the function we want to hack? We only need to dispatch on `non_dimensional_part`?
> 
> I added some suggestions to fix type stability, but LGTM aside from that

Correct. This is an attempt to minimize boilerplate for folks.

After the revision with Yair, I moved the non-dimensional functions into the `nondimensional_exchange_function.jl` file, which is where parts of the moisture deficit closure lived. 
We think this setup should work for 95% of people, and for the other 5% we can deal with that whenever problems arise.

yairchn reviewed (approved):
LGTM - needs rebase and update tables of course. great work I like this form of the code much more

yairchn commented:
Bors r+
